lua/apisix/plugins/wolf-rbac.lua

```
@@ -0,0 +1,386 @@
+                            ",body:", core.json.delay_encode(res.body))
+            if res.status < 500 then
+                break
+            else
+                core.log.info("request [curl -v ", url, "] failed! status:", res.status)
+                if i < retry_max then
+                    ngx.sleep(0.1)
+                end
+            end
+        end
+    end
+
+    if err then
+        core.log.error("fail request: ", url, ", err:", err)
+        return {status = 500, err = "request to wolf-server failed, err:" .. tostring(err)}
+    end
+
+    if res.status ~= 200 and res.status ~= 401 then
+        return {status = 500, err = 'request to wolf-server failed, status:' .. tostring(res.status)}
+    end
+
+    local body, err = json.decode(res.body)
+    if err then
+        errmsg = 'check permission failed! parse response json failed!'
+        core.log.error( "json.decode(", res.body, ") failed! err:", err)
+        return {status = res.status, err = errmsg}
+    else
+        if body.data then
+            userInfo = body.data.userInfo
+        end
+        errmsg = body.reason
+        return {status = res.status, err = errmsg, userInfo = userInfo}
+    end
+end
+
+
+function _M.rewrite(conf, ctx)
+    local url = ctx.var.uri
+    local action = ctx.var.request_method
+    local clientIP = core.request.get_ip(ctx)
+    local permItem = {action = action, url = url, clientIP = clientIP}
+    core.log.info("hit wolf-rbac rewrite")
+
+    local rbac_token = fetch_rbac_token(ctx)
+    if rbac_token == nil then
+        core.log.info("no permission to access ", core.json.delay_encode(permItem), ", need login!")
+        return 401, {message = "Missing rbac token in request"}
+    end
+
+    local tokenInfo = parse_rbac_token(rbac_token)
+    core.log.info("token info: ", core.json.delay_encode(tokenInfo))
+    if tokenInfo.err then
+        return 401, {message = 'invalid rbac token: parse failed'}
```

`{message = 'invalid rbac token: parse failed, ' .. err}`

Should we return the error message?

---

Done